sympy/combinatorics/fp_groups.py

```diff
@@ -0,0 +1,716 @@
+        self.table[alpha][self.A_dict[x]] = beta
+        self.table[beta][self.A_dict_inv[x]] = alpha
+
+    def define_f(self, alpha, x):
+        A = self.A
+        if self.table == CosetTableDefaultMaxLimit:
+            # abort the further generation of cosets
+            return
+        self.table.append([None]*len(A))
+        # beta is the new coset generated
+        beta = len(self.table) - 1
+        self.p.append(beta)
+        self.table[alpha][self.A_dict[x]] = beta
+        self.table[beta][self.A_dict_inv[x]] = alpha
+        # append to deduction stack
+        self.deduction_stack.append((alpha, x))
+
+    def scan_f(self, alpha, word):
+        """
+        >>> from sympy.combinatorics.free_group import free_group
+        >>> from sympy.combinatorics.fp_groups import CosetTable, FpGroup
+        >>> F, x, y = free_group("x, y")
+        >>> f = FpGroup(F, [x**3, y**3, x**-1*y**-1*x*y])
+        >>> c = CosetTable(f, [x])
+
+        """
+        # alpha is an integer representing a "coset"
+        # since scanning can be in two cases
+        # 1. for alpha=0 and w in Y (i.e generating set of H)
+        # 2. alpha in omega (set of live cosets), w in R (relators)
+        A_dict = self.A_dict
+        A_dict_inv = self.A_dict_inv
+        f = alpha
+        i = 0
+        r = len(word)
+        b = alpha
+        j = r - 1
+        # list of union of generators and their inverses
+        while i <= j and self.table[f][A_dict[word[i]]] is not None:
+            f = self.table[f][A_dict[word[i]]]
+            i += 1
+        # can this be replaced with i == r ?
+        if i > j:
+            if f != b:
+                # implement the "coincidence" routine on Pg 158 of Handbook.
+                self.coincidence_f(f, b)
+            return
+        while j >= i and self.table[b][A_dict_inv[word[j]]] is not None:
+            b = self.table[b][A_dict_inv[word[j]]]
+            j -= 1
+        if j < i:
+            # we have an incorrect completed scan with coincidence f ~ b
+            # run the "coincidence" routine
+            self.coincidence_f(f, b)
+        elif j == i:
+            # deduction process
+            self.table[f][A_dict[word[i]]] = b
+            self.table[b][A_dict_inv[word[i]]] = f
+            self.deduction_stack.append((f, word[i]))
+        # otherwise scan is incomplete and yields no information
+
+    # alpha, beta coincide
+    # here alpha, beta represent the pair of cosets where
+    # coincidence occurs
+    def coincidence_f(self, alpha, beta):
+        """
+        """
+        A_dict = self.A_dict
+        A_dict_inv = self.A_dict_inv
+        p = self.p
+        l = 0
+        # behaves as a queue
+        q = []
+        self.merge(alpha, beta, q)
+        while len(q) > 0:
+            gamma = q.pop(0)
+            # comment by Kalevi, this is already done by p[v] = mu
+            # del C[gamma]
+            # commenting this out
+            # omega = omega - set(gamma)
```

Comment:
I think these comments could be dropped now. It seems that the deletion was some kind of typo in the original pseudocode.